hs-bindgen-runtime/src/HsBindgen/Runtime/FunPtr/Common.hs

```diff
@@ -0,0 +1,172 @@
+-- to C function pointers.
+--
+-- This module provides a type class 'ToFunPtr' that allows for a uniform
+-- interface to convert Haskell functions to C function pointers.
+--
+-- This module also provides TH splices that generate FFI wrappers and
+-- 'ToFunPtr' instances that are called in different modules to paralellize
+-- compilation and compile time code generation.
+--
+module HsBindgen.Runtime.FunPtr.Common (
+    -- * Type class
+    ToFunPtr(..)
+  , FromFunPtr(..)
+
+    -- * Template Haskell FFI wrapper generation and 'ToFunPtr' instance
+    -- generation
+
+    -- ** Types
+  , allPrimTypes
+  , allPtrTypes
+  , allIOTypes
+    -- *** Common Types
+  , commonPrimTypes
+  , commonPtrTypes
+  , commonReturnTypes
+    -- ** Generate the instances
+  , generateInstance
+  ) where
+
+import Data.Void (Void)
+import Foreign qualified as F
+import Foreign.C.Types
+import GHC.Ptr qualified as Ptr
+import Language.Haskell.TH
+
+-- | Type class for converting Haskell functions to C function pointers.
+--
+class ToFunPtr a where
+  -- | Convert a Haskell function to a C function pointer.
+  --
+  -- The caller is responsible for freeing the function pointer using
+  -- 'F.freeHaskellFunPtr' when it is no longer needed.
+  --
+  toFunPtr :: a -> IO (F.FunPtr a)
+
+-- | Type class for converting C function pointers to Haskell functions.
+--
+class FromFunPtr a where
+  -- | Convert C function pointer into a Haskell function.
+  fromFunPtr :: F.FunPtr a -> a
+
+-- | Get primitive marshallable types
+--
+allPrimTypes :: [Q Type]
+allPrimTypes =
+  [ [t| CChar      |]
+  , [t| CSChar     |]
+  , [t| CUChar     |]
+  , [t| CInt       |]
+  , [t| CUInt      |]
+  , [t| CShort     |]
+  , [t| CUShort    |]
+  , [t| CLong      |]
+  , [t| CULong     |]
+  , [t| CPtrdiff   |]
+  , [t| CSize      |]
+  , [t| CLLong     |]
+  , [t| CULLong    |]
+  , [t| CBool      |]
+  , [t| CFloat     |]
+  , [t| CDouble    |]
+  , [t| Int        |]
+  ]
+
+-- | Get pointer types for all primitive types
+--
+allPtrTypes :: [Q Type]
+allPtrTypes = [t| Ptr.Ptr Void |]
+            : [ [t| Ptr.Ptr $t |]
+              | t <- allPrimTypes
+              ]
+
+-- | Get IO types for all primitive types
+--
+allIOTypes :: [Q Type]
+allIOTypes = [t| IO () |]
+           : [ [t| IO $t |]
+             | t <- allPrimTypes
+             ]
+
+-- | A list of the most common primitive types found in C callbacks.
+--
+commonPrimTypes :: [Q Type]
+commonPrimTypes =
+  [ [t| CChar   |]
+  , [t| CInt    |]
+  , [t| CUInt   |]
+  , [t| CDouble |]
+  , [t| CFloat  |]
+  ]
+
+-- | Common pointer types, including the crucial @void*@ and @char*@.
+--
+commonPtrTypes :: [Q Type]
+commonPtrTypes =
+  [ [t| Ptr.Ptr Void  |]
+  , [t| Ptr.Ptr CChar |]
+  , [t| Ptr.Ptr CInt  |]
+  ]
+
+-- | The most common return types for callbacks are @void@ and integer status
+-- codes.
+--
+commonReturnTypes :: [Q Type]
+commonReturnTypes =
+  [ [t| IO ()   |]
+  , [t| IO CInt |]
+  ]
```

Comment:
Should this be called `commonIOTypes` to be consistent with `allIOTypes`? Or should `allIOTypes` be called `allReturnTypes`?

Reply:
That's a good question but, I think the two names should be different, one of them is used for generating nullary functions (0 argument function) instances and the other is used for functions that take arguments.